dsgrid/query/derived_dataset.py

```diff
@@ -0,0 +1,178 @@
+
+import json5
+
+from dsgrid.config.dataset_config import (
+    DataClassificationType,
+    DataSchemaType,
+    InputDatasetType,
+    ColumnType,
+)
+from dsgrid.data_models import serialize_model
+from dsgrid.utils.spark import read_dataframe, get_unique_values
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.query.models import ProjectQueryModel, DatasetMetadataModel
+from dsgrid.query.query_submitter import QuerySubmitterBase
+from dsgrid.registry.dataset_registry import DatasetRegistry
+from dsgrid.registry.registry_manager import RegistryManager
+
+
+logger = logging.getLogger(__name__)
+
+
+def create_derived_dataset_confg_from_query(
+    query_path: Path, dst_path: Path, registry_manager: RegistryManager
+):
+    """Create a DatasetConfigModel and dimensions from a query result.
+
+    Parameters
+    ----------
+    query_path : Path
+        Output directory from a query.
+    dst_path : Path
+        Directory in which to create the dataset config files.
+    registry_manager : RegistryManager
+
+    Returns
+    -------
+    bool
+        Returns True if the operation is successful.
+    """
+    metadata_file = QuerySubmitterBase.metadata_filename(query_path)
+    query_file = QuerySubmitterBase.query_filename(query_path)
+    table_file = QuerySubmitterBase.table_filename(query_path)
+    if not metadata_file.exists() or not query_file.exists():
+        logger.error("%s is not a valid query result directory", query_path)
+        return False
+
+    query = ProjectQueryModel.from_file(query_file)
+    if not does_query_support_a_derived_dataset(query):
+        return False
+
+    metadata = DatasetMetadataModel.from_file(metadata_file)
+    project = registry_manager.project_manager.load_project(
+        query.project.project_id, version=query.project.version
+    )
+    # TODO: should there be a warning if the current project version is later?
+    dimensions = []
+    dimension_references = []
+    for dim_type in DimensionType:
+        dims = getattr(metadata.dimensions, dim_type.value)
+        assert len(dims) == 1, dims
+        dim = project.config.get_dimension(next(iter(dims)))
+        df = read_dataframe(table_file)
+        if metadata.pivoted.dimension_type == dim_type:
+            is_valid = bool(metadata.pivoted.columns.difference(df.columns))
+            table_records = metadata.pivoted.columns
+        elif dim_type == DimensionType.TIME:
+            is_valid = True  # TODO: how can time be different? This was already mapped to project
+            table_records = None
+        else:
+            is_valid = is_dimension_valid_for_dataset(dim, df)
+            table_records = get_unique_values(df, dim.model.dimension_query_name)
+        if is_valid:
+            dim_ref = project.config.get_dimension_reference(dim.model.dimension_id)
+            dimension_references.append(serialize_model(dim_ref))
+        else:
+            records = dim.get_records_dataframe()
```

Comment:
This is a problematic section that may need changes. Suppose we are generating the dataset config for comstock_projected_2022. The dataframe produced by the query has project-mapped dimensions. The project dimension records for sector and metric do not match the records in the dataframe. The easiest and safest thing to do is generate a new records file for each of these dimensions and register a new dimension (that's what this code does). However, as it happens, this new records file will be identical to what the original comstock_reference dataset has. That obviously won't always be the case because a dataset may use different record IDs than the project.

So, what do we do here for dimensions that are not identical to the project?
1. Always generate new ones, even if duplicate. Rely on the user to change name and description so that they are not misleading.
2. Identify the source datasets from the query, read their dimension records and check for a match. Use the first dimension that matches. This would mean taking name, description, display_name, and query_name regardless of whether they are appropriate.
3. Loop through all dimensions looking for a match in records. Take the first one that matches.

Reply:
This is pertinent to the general registration of datasets and whether existing dim records should be reused as much as possible.

This will get complicated more in terms of version control. I wanna say dimensions should be reused as much as possible within a project? So option 2 or 3 (within project)?